keycloak_client_rolescope: fix for issue #10342

killianlevacher opened the pull request:
##### SUMMARY

this is a fix for issue #10342


Changelog Fragment:

This enables the `community.general.keycloak_client_rolescope` Ansible task to assign roles to client scopes in Keycloak, which currently fails because of a bug described in #10342





##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
community.general.keycloak_client_rolescope

##### ADDITIONAL INFORMATION
Steps to reproduce the bug described in details in #10342

killianlevacher committed:
fix for issue #10342

desand01 commented:
Hi @killianlevache, thank you for your contribution.

It seems that the integration tests are currently failing. To help find a solution that addresses your use case without breaking compatibility, would you mind adding an integration test that reproduces the scenario you are trying to support?

This would help me better understand the context.

Thanks again!

### general/tests/integration/targets/keycloak_client_rolescope
TASK [keycloak_client_rolescope : Map roles to public client] ******************
fatal: [testhost]: FAILED! => {"changed": false, "msg": "Failed to retrieve client 'myrealm.backend-client-private'"}

flvmz commented:
I found this yesterday and commented on the bug report, because I think this module is *not* ment to manage roles in the context of `client scopes`: https://github.com/ansible-collections/community.general/issues/10342#issuecomment-3066717673

killianlevacher committed:
reverted suggested fixes to original code and added example where integration test would fail

killianlevacher commented:
@desand01 Apologies for only coming back to you now on this. It took me a while to dig into what was going on. 

For the purpose of this conversation, I commented out my suggested fixes and reverted to the original code to demonstrate the task failing. I also added steps in the integration test `main.yml` to demonstrate where the step fails.

As pointed out in `main.yml` the step which attempts to assign a role to a client scope fails under the original code.
When client_scope_id is given instead the {{ client_name_private }} (as for the task 'Map roles to public client') this step does not fail, however no role is assigned to this scope

With the fix I am suggesting, when client_scope_id is provided with the scope_id (as opposed to a client name), the role is indeed assigned to the client scope as expected but the 'Map roles to public client' task fails probably for the reverse reasons (haven't been able to figure out why yet)

Unless mistaken, somehow the underlying code seems to be confusing the notion of a client_scope_id with that of a client_id.

I agree with @flvmz  that it seems like this mapping should be in the client_scope module to start with. 

Perhaps, something which could clarify the confusion going on. Can someone explain, within the task called ` Map roles to public client` in the integration test `main.yml` why are we giving a client id to a parameter called `client_scope_id`  `client_scope_id: "{{ client_name_private }}"`? This seems to be the root of the problem/confusion.

desand01 commented:
Hi @killianlevacher,
Apologies for the delay in getting back to you.

Unfortunately, I think there is currently no module in the collection that allows you to manage client scopes at the realm level.

As a workaround, you can use the keycloak_client_rolescope module to assign role scopes of another client directly to a client. However, this solution has some limitations—for example, it does not allow you to distinguish between default and optional scopes when assigning them.

If your use case requires managing reusable client scope templates at the realm level and assigning them to clients (with support for default/optional designation), I suggest opening feature requests (Issues) for two new modules (names are only suggestions):

keycloak_realm_client_scope: to manage client scope definitions and their role mappings at the realm level.
keycloak_realm_client_scopemapping: to handle the assignment of those scopes (as default or optional) to individual clients.

desand01 commented:
Hi @killianlevacher, maybe this is what you are looking for:
community.general.keycloak_clientscope: create your custom realm scope
community.general.keycloak_clientscope_type: assign the realm scope to a client, either Default or Optional

felixfontein commented:
How should we continue with this PR?

needs_info

desand01 commented:
Hi,

I think this PR should be closed.
The required functionality is already available through existing modules:
community.general.keycloak_clientscope: create your custom realm scope
community.general.keycloak_clientscope_type: assign the realm scope to a client (either Default or Optional)
community.general.keycloak_client: as of version 11.4.0, use default_client_scopes or optional_client_scopes to assign realm scopes to a client
community.general.keycloak_client_rolescope: directly grant scope from Client A role to Client B
